File: 2691-count-vowel-strings-in-ranges/count-vowel-strings-in-ranges.js
/**
 * @param {string[]} words
 * @param {number[][]} queries
 * @return {number[]}
 */
var vowelStrings = function (words, queries) {
    const n = words.length;
    const vowels = new Set(['a', 'e', 'i', 'o', 'u']);

    const dp = Array(n).fill(0);
    for (let i = 0; i < n; i++) {
        if (vowels.has(words[i][0]) && vowels.has(words[i][words[i].length - 1])) {
            dp[i] = 1;
        }
    }

    const prefixSum = Array(n + 1).fill(0);
    for (let i = 0; i < n; i++) {
        prefixSum[i + 1] = prefixSum[i] + dp[i];
    }

    const result = [];
    for (const [start, end] of queries) {
        result.push(prefixSum[end + 1] - prefixSum[start]);
    }

    return result;
};

File: 2691-count-vowel-strings-in-ranges/count-vowel-strings-in-ranges.py
class Solution:
    def vowelStrings(self, words: List[str], queries: List[List[int]]) -> List[int]:
        n = len(words)
        vowel = "aeiou"

        dp = [1 if word[0] in vowel and word[-1] in vowel else 0 for word in words]

        prefixsum = [0] * (n + 1)
        for i in range(n):
            prefixsum[i + 1] = prefixsum[i] + dp[i]

        ans = []

        for start, end in queries:
            count = prefixsum[end + 1] - prefixsum[start]
            ans.append(count)

        return ans
